ownCloud/Settings/Passcode/PasscodeManager.swift

```diff
@@ -0,0 +1,375 @@
+        self.window?.rootViewController = self.passcodeViewController!
+        self.window?.makeKeyAndVisible()
+    }
+
+    func homeButtonPressed() {
+
+        //Protection to hide the PasscodeViewController if is on Delete modes to show the Unlock
+        if self.passcodeViewController != nil {
+            if self.passcodeMode == .deletePasscode ||
+                self.passcodeMode == .deletePasscodeError {
+                self.completionHandler!()
+                self.window?.isHidden = true
+                self.passcodeViewController = nil
+            }
+        }
+
+        //Store the date when the user pressed home button
+        if self.isPasscodeActivated, self.datePressedHomeButton == nil {
+            self.datePressedHomeButton = Date()
+        }
+
+        //Show the passcode
+        self.showPasscodeIfNeeded(hiddenOverlay: false)
+    }
+
+    // MARK: - Interface updates
+
+    private func updateUI() {
+
+        var messageText : String?
+        var errorText : String? = ""
+
+        switch self.passcodeMode {
+        case .addPasscodeFirstStep?:
+            messageText = "Enter code".localized
+
+        case .addPasscodeSecondStep?:
+            messageText = "Repeat code".localized
+
+        case .unlockPasscode?:
+            messageText = "Enter code".localized
+            self.passcodeViewController?.cancelButton?.isHidden = true
+
+        case .unlockPasscodeError?:
+            messageText = "Enter code".localized
+            errorText = "Incorrect code".localized
+            self.passcodeViewController?.cancelButton?.isHidden = true
+
+        case .deletePasscode?:
+            messageText = "Delete code".localized
+
+        case .deletePasscodeError?:
+            messageText = "Delete code".localized
+            errorText = "Incorrect code".localized
+
+        case .addPasscodeFirstStepAfterErrorOnSecond?:
+            messageText = "Enter code".localized
+            errorText = "The entered codes are different".localized
+
+        default:
+            break
+        }
+
+        self.passcodeViewController?.passcodeValueTextField?.text = ""
+        self.passcodeViewController?.messageLabel?.text = messageText
+        self.passcodeViewController?.errorMessageLabel?.text = errorText
+        self.passcodeViewController?.timeTryAgainMessageLabel?.text = ""
+    }
+
+    func dismissAskedPasscodeIfDateToAskIsLower() {
+
+        if shouldBeLocked {
+            self.passcodeViewController?.hideOverlay()
+        } else {
+            if self.passcodeViewController != nil {
+                //Protection to hide the PasscodeViewController only if is in unlock mode
+                if self.passcodeMode == .unlockPasscode ||
+                    self.passcodeMode == .unlockPasscodeError {
+                    self.dismissPasscode()
+                    self.datePressedHomeButton = nil
+                }
+            }
+        }
+    }
+
+    func cancelButtonTaped() {
```

Comment:
if you follow my suggestion of coding the `completionHandler` in the `dismissPasscode()` method, you can get rid of this and call `dismissPasscode` when the cancel button is pressed.

Reply:
Yep, I included this on the same method